### PR TITLE
Statistical testing

### PR DESCRIPTION
**Pull request by etwest**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -251,6 +251,36 @@ else ()
...
+  test/util/graph_gen.cpp
+  src/graph.cpp
+  src/supernode.cpp
+  src/graph_worker.cpp
+  src/work_queue.cpp
+  src/l0_sampling/sketch.cpp
+  src/l0_sampling/update.cpp
+  src/util.cpp)
+target_link_libraries(statistical_test PRIVATE xxHash::xxhash GTest::gtest)
+target_compile_definitions(statistical_test PRIVATE VERIFY_SAMPLES_F)
```
Comment (Victor-C-Zhang) [inline CMakeLists.txt]: if this is running as our cron job, shouldn't we have both with and without gutter tree?

File: test/statistical_testing/graph_testing.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    return failures;
+}
+
+int medium_with_iso_test(int runs) {
+    int failures = 0;
+    for (int i = 0; i < runs; i++) {
+        generate_stream({2048,0.002,0.5,0,"./sample.txt","./cumul_sample.txt"});
+        std::fstream graph_file{"./sample.txt"};
+        std::fstream cumul_file{"./cumul_sample.txt"};
+        graph_file.write("2070", 4); // increase the node size to create iso nodes
```
Comment (Victor-C-Zhang) [inline test/statistical_testing/graph_testing.cpp:52]: what are iso nodes?

File: test/statistical_testing/analyze_results.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+
+import numpy as np
+import argparse
+from scipy.stats import ttest_ind
+
+def check_error(test_name, test_result_file, expected_result_file, confidence=0.95):
```
Comment (Victor-C-Zhang) [inline test/statistical_testing/analyze_results.py:6]: why do you take a confidence argument if you never use it?

File: test/statistical_testing/graph_testing.cpp (hunk 2 above)
Comment (etwest) [reply test/statistical_testing/graph_testing.cpp:52]: So this is basically a test which approximates our kron streams where there are a few nodes which are guaranteed to not have any neighbors.

File: test/statistical_testing/analyze_results.py (hunk 3 above)
Comment (etwest) [reply test/statistical_testing/analyze_results.py:6]: I want to use it but I don't know how to determine the z-value from the confidence value. Do you?

File: CMakeLists.txt (hunk 1 above)
Comment (etwest) [reply CMakeLists.txt]: Yes I suppose so, I'll implement that change

File: test/statistical_testing/analyze_results.py (hunk 3 above)
Comment (Victor-C-Zhang) [reply test/statistical_testing/analyze_results.py:6]: Suppose our confidence is 2c. Then we want to find the value of x such that the Normal CDF takes value .5+c

The CDF is given
.5[1 + erf(x/sqrt2)]

You can find this x value by any means, for instance Newton or even a dumb binary search.
I can implement this tonight if you’d like.

**Comment by etwest** [comment]
Thanks Victor

File: CMakeLists.txt (hunk 1 above)
Comment (etwest) [reply CMakeLists.txt]: Done

**Review by Victor-C-Zhang — APPROVED** [review]
Approving changes, but still worried about test failure rates

**Pull request merged**